api/utils/keys/piv/service.go

```diff
@@ -52,7 +57,111 @@ func NewYubiKeyService(prompt hardwarekey.Prompt) *YubiKeyService {
 //   - touch  & pin  -> 9d
 //   - touch  & !pin -> 9e
 func (s *YubiKeyService) NewPrivateKey(ctx context.Context, config hardwarekey.PrivateKeyConfig) (*hardwarekey.PrivateKey, error) {
-	return newPrivateKey(ctx, config, s.prompt, s)
+	// Get the requested or default PIV slot.
+	var slotKey hardwarekey.PIVSlotKey
+	var err error
+	if config.CustomSlot != "" {
+		slotKey, err = config.CustomSlot.Parse()
+	} else {
+		slotKey, err = hardwarekey.GetDefaultSlotKey(config.Policy)
+	}
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	pivSlot, err := parsePIVSlot(slotKey)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// If the program has already retrieved and cached this key, return it.
+	if key, ok := cachedKeys[pivSlot]; ok && key.GetPromptPolicy() == config.Policy {
+		return key, nil
+	}
+
+	// Use the first yubiKey we find.
+	y, err := FindYubiKey(0, s.prompt)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// If PIN is required, check that PIN and PUK are not the defaults.
+	if config.Policy.PINRequired {
+		if err := y.checkOrSetPIN(ctx); err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	generatePrivateKey := func() (*hardwarekey.PrivateKey, error) {
+		ref, err := y.generatePrivateKey(pivSlot, config.Policy)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return hardwarekey.NewPrivateKey(s, ref), nil
+	}
+
+	// If a custom slot was not specified, check for a key in the
+	// default slot for the given policy and generate a new one if needed.
+	if config.CustomSlot == "" {
+		switch cert, err := y.getCertificate(pivSlot); {
+		case errors.Is(err, piv.ErrNotFound):
+			return generatePrivateKey()
+
+		case err != nil:
+			return nil, trace.Wrap(err)
+
+		// Unknown cert found, this slot could be in use by a non-teleport client.
+		// Prompt the user before we overwrite the slot.
+		case len(cert.Subject.Organization) == 0 || cert.Subject.Organization[0] != certOrgName:
+			if err := s.promptOverwriteSlot(ctx, nonTeleportCertificateMessage(pivSlot, cert)); err != nil {
+				return nil, trace.Wrap(err)
+			}
+			return generatePrivateKey()
+		}
+	}
+
+	// Check for an existing key in the slot that satisfies the required
+	// prompt policy, or generate a new one if needed.
+	slotCert, attCert, att, err := y.attestKey(pivSlot)
+	switch {
+	case errors.Is(err, piv.ErrNotFound):
+		return generatePrivateKey()
+
+	case err != nil:
+		return nil, trace.Wrap(err)
+
+	case config.Policy.TouchRequired && att.TouchPolicy == piv.TouchPolicyNever:
```

Comment:
I know this is moved code, so please treat this as a side question: what does it exactly mean that the config policy says "touch required", and the attestation's policy says "touch never required"? Where does this configuration come from? (I was scratching my head, thinking why this situation is allowed at all.)

Reply:
It's possible for a key to be pre-generated in the slot that the Teleport Client wants to use to satisfy a specific key policy (`hardware_key`|`hardware_key_touch`, etc...). For example, it's possible for a user to generate a new key on their YubiKey with `ykman piv keys generate 9c --touch-policy=never ...`. `9c` is the PIV slot that Teleport uses for `hardware_key_touch` by default, so the Teleport client needs to detect that an unsatisfying key has been pre-generated in that slot, and needs to be overwritten to continue to login.